IdentityCore/src/webview/embeddedWebview/challangeHandlers/MSIDClientTLSHandler.m

```diff
@@ -0,0 +1,234 @@
+        //
+        // Using DefaultHandling will result in the OS not handing back client TLS
+        // challenges for another ~60 seconds, behavior that looks broken in the
+        // user CBA case, but here is masked by the user having to enroll their
+        // device.
+        completionHandler(NSURLSessionAuthChallengePerformDefaultHandling, nil);
+        return YES;
+    }
+    
+    MSID_LOG_INFO(context, @"Responding to WPJ cert challenge");
+    MSID_LOG_INFO_PII(context, @"Responding to WPJ cert challenge. host: %@", challenge.protectionSpace.host);
+    
+    NSURLCredential *creds = [NSURLCredential credentialWithIdentity:info.securityIdentity
+                                                        certificates:@[(__bridge id)info.certificate]
+                                                         persistence:NSURLCredentialPersistenceNone];
+    
+    completionHandler(NSURLSessionAuthChallengeUseCredential, creds);
+    
+    return YES;
+}
+
+#if !TARGET_OS_IPHONE
+
++ (BOOL)handleCertAuthChallenge:(NSURLAuthenticationChallenge *)challenge
+                        webview:(WKWebView *)webview
+                        context:(id<MSIDRequestContext>)context
+              completionHandler:(ChallengeCompletionHandler)completionHandler
+{
+    NSString *host = challenge.protectionSpace.host;
+    NSArray<NSData*> *distinguishedNames = challenge.protectionSpace.distinguishedNames;
+    
+    // Otherwise check if a preferred identity is set for this host
+    SecIdentityRef identity = SecIdentityCopyPreferred((CFStringRef)host, NULL, (CFArrayRef)distinguishedNames);
+    
+    if (!identity)
+    {
+        // If there was no identity matched for the exact host, try to match by URL
+        // URL matching is more flexible, as it's doing a wildcard matching for different subdomains
+        // However, we need to do both, because if there's an entry by hostname, matching by URL won't find it
+        identity = SecIdentityCopyPreferred((CFStringRef)webview.URL.absoluteString, NULL, (CFArrayRef)distinguishedNames);
+    }
+    
+    if (identity != NULL)
+    {
+        MSID_LOG_INFO(context, @"Using preferred identity");
+        [self respondCertAuthChallengeWithIdentity:identity context:context completionHandler:completionHandler];
+    }
+    else
+    {
+        // If not prompt the user to select an identity
+        [self promptUserForIdentity:distinguishedNames
+                               host:host
+                            webview:webview
+                      correlationId:context.correlationId
+                  completionHandler:^(SecIdentityRef identity)
+         {
+             if (identity == NULL)
+             {
+                 MSID_LOG_INFO(context, @"No identity returned from cert chooser");
+                 
+                 // If no identity comes back then we can't handle the request
+                 completionHandler(NSURLSessionAuthChallengeRejectProtectionSpace, nil);
```

Comment:
what will happen with loading in this case?

Reply:
I have tested two different accounts which have two different behaviors:

For one account (that appears in one of our IcMs), after we reject the TLS challenge, the page continues to load,`webView:didFinishNavigation:` get called and spinner is dismissed;
For another account (the jason CBA test account), after we reject the TLS challenge, the page doesn't continue to load, no delegate gets called and spinner is not dismissed.

So I guess it might depend on server for what will happen if we reject the TLS challenge.
I have asked for help from Apple Support but haven't heard back yet.

Opened the following issue to track if we could improve the experience:
https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/issues/156